From: =?utf-8?b?Ik1hcmNvIFRyZXZpc2FuIChUcmV2acOxbyki?= <mail@3v1n0.net>
Date: Thu, 16 Apr 2026 04:33:45 +0200
Subject: appFavorites: Only emit :changed signal on changes

We were emitting the :changed signal on every setting or app-filter
change even though the actual favorites list might not have changed.

So now every time it's relevant, check if something is changed before
emitting the :changed signal.

We do a deep check of the app IDs as potentially an app could have
been removed while another is being installed during the same update
window, and the length of the IDs list wouldn't change.

Origin: https://gitlab.gnome.org/GNOME/gnome-shell/-/merge_requests/4180
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/gnome-shell/+bug/2148339
---
 js/ui/appFavorites.js | 28 +++++++++++++++++++++-------
 1 file changed, 21 insertions(+), 7 deletions(-)

diff --git a/js/ui/appFavorites.js b/js/ui/appFavorites.js
index b4dec3a..a88f98c 100644
--- a/js/ui/appFavorites.js
+++ b/js/ui/appFavorites.js
@@ -9,20 +9,34 @@ class AppFavorites extends Signals.EventEmitter {
 
         // Filter the apps through the user’s parental controls.
         this._parentalControlsManager = ParentalControlsManager.getDefault();
-        this._parentalControlsManager.connect('app-filter-changed', () => {
-            this.reload();
-            this.emit('changed');
-        });
+        this._parentalControlsManager.connect('app-filter-changed', () =>
+            this._updateFavorites());
 
         this.FAVORITE_APPS_KEY = 'favorite-apps';
         this._favorites = {};
-        global.settings.connect(`changed::${this.FAVORITE_APPS_KEY}`, this._onFavsChanged.bind(this));
+        global.settings.connect(`changed::${this.FAVORITE_APPS_KEY}`, () =>
+            this._updateFavorites({checkOrder: true}));
+
         this.reload();
     }
 
-    _onFavsChanged() {
+    _updateFavorites(params = {checkOrder: false}) {
+        const oldIDs = this._getIds();
         this.reload();
-        this.emit('changed');
+        const newIDs = this._getIds();
+
+        const haveChanged = () => {
+            if (oldIDs.length !== newIDs.length)
+                return true;
+
+            if (params.checkOrder)
+                return !oldIDs.every((id, i) => id === newIDs[i]);
+
+            return oldIDs.some(id => !this.isFavorite(id));
+        };
+
+        if (haveChanged())
+            this.emit('changed');
     }
 
     reload() {
